### ISSUE TITLE
Unique coming back to life?

### ISSUE THREAD

**Issue by angband-import** [angband/angband#4245]
**Reported by NickMcConnell on 27 Dec 2019 09:56 UTC**
See http://angband.oook.cz/forum/showpost.php?p=141990&postcount=141

**Comment by NickMcConnell** [comment]
[Quaker](http://angband.oook.cz/forum/showpost.php?p=141990&postcount=141) and [Beorn](http://angband.oook.cz/forum/showpost.php?p=143264&postcount=194) are both suspected of this. The fact that monster knowledge showed them as dead and lack of any indication of shapechange indicates that it's probably not just other monsters pretending to be them.

**Comment by NickMcConnell** [comment]
I haven't been able to reproduce this, maybe 4c34efd has fixed it.  I have tightened up the code around checking for uniques' death a little too, so we'll just have to see how this goes.

**Comment by NickMcConnell** [comment]
Nope, still happens, Quaker again.

**Comment by backwardsEric** [comment]
Reviewing the code related to shape changes and monster death turned up three things that seemed suspicious to me, but none of them explain a unique coming back to life:

- The check for trampling of uniques in mon-move.c's monster_can_kill() only checks the current race; should it also check the original race?  Perhaps add "|| (mon1->original_race && rf_has(mon1->original_race->flags, RF_UNIQUE))" to the current check?
- When setting original_race in mon-util.c's monster_change_shape(), should it check that original_race is already not NULL to account for something that changed shape into something that also has the ability to change shape? That would replace "mon->original_race = mon->race;" with "if (!mon->original_race) mon->original_race = mon->race;".  However, I didn't see any monsters in the current list that could chain shape changes like that.
- When choosing a shape to change into using the explicitly listed shapes, there's no check for the case if the shape is a unique race which is already present on the level.  However, I didn't see anything that would go wrong if that occurred (Quaker present + Maia changes shape to Quaker; Huan present + Maia changes shape to Huan).

**Comment by NickMcConnell** [comment]
Agreed point 3 is OK (and possibly desirable), but I think the other two changes are worth making.

**Comment by NickMcConnell** [comment]
Setting this to Ongoing.
